File: haskell/027/Main.hs
{-
    ProjectEuler 027
    Haskell / Main.hs
    copyright (c) 2015 Susisu

    GHC 7.8.3
    `ghc -O2 Main.hs -o bin/main`
-}

module Main where

import Data.List

main :: IO ()
main = print . fst $ foldl1' (\x@(_, r) y@(_, s) -> if r > s then x else y) results
    where
        formulas = [(a * b, formula a b) | a <- [-1000 .. 1000], b <- [-1000 .. 1000]]
        results  = map (\(p, f) -> (p, calc f)) formulas

calc :: (Int -> Int) -> Int
calc f = itr 0 where itr n = if isPrime (f n) then itr (n + 1) else n

formula :: Int -> Int -> Int -> Int
formula a b n = n ^ 2 + a * n + b

isPrime :: Int -> Bool
isPrime n
    | n <= 1    = False
    | otherwise = itr 2
    where
        r :: Int
        r = floor . sqrt $ fromIntegral n
        itr :: Int -> Bool
        itr m
            | m <= r =
                if n `mod` m == 0 then
                    False
                else
                    itr (m + 1)
            | otherwise = True

File: ocaml/027/main.ml
(*
    ProjectEuler 027
    OCaml / main.ml
    copyright (c) 2016 Susisu

    OCaml 4.02.3
    `ocamlfind ocamlopt -o bin/main main.ml`
*)

let is_prime n =
    let rec itr n m =
        if float_of_int m > sqrt (float_of_int n) then true
        else if n mod m = 0 then false
        else itr n (m + 1)
    in
    if n < 2 then false
    else itr n 2

let seq a b =
    let rec itr c n =
        let p = n * n + a * n + b in
        if is_prime p then itr (c + 1) (n + 1)
        else c
    in itr 0 0

let main _ =
    let rec itr m ma mb a b =
        if a > 1000 then ma * mb
        else if b > 1000 then itr m ma mb (a + 1) 2
        else
            let c = seq a b in
            if c > m then itr c a b a (b + 1)
            else itr m ma mb a (b + 1)
    in
    let m = itr 0 0 0 (-1000) 2 in
    print_int m;
    print_newline ()

let _ = main ()
